checkbox-provider-ce-oem/bin/socketcan_test.py

```diff
@@ -0,0 +1,325 @@
+
+        if recv_id != can_id_i or recv_data != data_b:
+            raise SystemExit('ERROR: ID/Data received does not match sent')
+        else:
+            logging.info('\nPASSED')
+
+
+def stress_echo_test(interface, can_id, eff_flag, fd_mode, count=30):
+    """
+    Perform the echo stress test through SocketCAN bus
+
+    Args:
+        interface (str): network interface of SocketCAN, e.g. can0
+        can_id (str):       CAN ID
+        eff_flag (bool):    EFF CAN ID Enable
+        fd_mode (bool):     FD mode enable
+        count (int):        Stress test cycle
+
+    Raises:
+        SystemExit: if received CAN frames is not expected
+    """
+    # Due to the timestamp data is 8 bytes
+    # So we need to generate extra random data (56 bytes)
+    data_size = 56 if fd_mode else 0
+
+    can_id_i, prefix_data, id_flags = _random_can_data(
+                        can_id, eff_flag, data_size)
+
+    logging.info("Initial CAN Link object with %s", interface)
+    with prepare_can_link(interface, fd_mode):
+        can_socket = CANSocket(interface, fd_mode)
+
+        logging.info("Generate data for stress tests")
+        original_records = [
+            can_socket.struct_packet(
+                can_id_i,
+                b"".join([
+                    prefix_data,
+                    bytes.fromhex(
+                        str(int(datetime.datetime.now().timestamp()*1000000)))
+                ]),
+                id_flags,
+                fd_frame=fd_mode
+            ) for _ in range(count)
+        ]
+
+        time_format = "%Y-%m-%d %H:%M:%S"
+        recv_records = []
+
+        start_time = datetime.datetime.now()
+        logging.info(
+            "# Start stress echo string test at %s",
+            start_time.strftime(time_format)
+        )
+        for can_pkt in original_records:
+            can_socket.send(can_pkt, timeout=5)
+            recv_records.append(can_socket.recv(5))
+            if recv_records[-1] is None:
+                logging.error(
+                    "Stop testing due to failed to receive packet"
+                )
+                logging.error(
+                    "Received %d packets from CAN echo server",
+                    len(recv_records)
+                )
+                raise SystemExit("CAN ECHO Stress test failed")
+
+        end_time = datetime.datetime.now()
+        logging.info("# End stress echo string test at {}".format(
+            end_time.strftime(time_format)
+        ))
+
+        elapsed_time = end_time - start_time
+        logging.info(
+            "received %s frames in %s", len(recv_records), elapsed_time)
+
+        if len(recv_records) == count:
+            logging.info("# Checking the data in received frames")
+            failed_count = 0
+            for index, data in enumerate(original_records):
+                # validate data field in CAN packet only
+                if can_socket.destruct_packet(recv_records[index]) != \
```

Comment:
I think we can add a helper function to compare the sent and received data like below.

``` python
def is_same_packet(sent_packet, received_packet):
    s_data = can_socket.destruct_packet(sent_packet)
    r_data = can_socket.destruct_packet(received_packet)
    is_same_data = s_data == r_data
    if not is_same_data:
        logging.error("Received packet is not same as Sent packet")
        logging.error("Received packet: %s", s_data)
        logging.error("Sent data: %s", r_data)
    return is_same_data
```

So we can modify the following section
```python
if is_same_packet(data, recv_records[index]):
    failed_count += 1
```

And we can also replace the comparison at Line 143

Reply:
It would help us to implement the unit test in the future

Reply:
fixed in [0b39c1d](https://github.com/canonical/checkbox-provider-ce-oem/pull/62/commits/0b39c1d9d9c8e6a6ecc9574bfc6f47fe60cfd039)